utils/s3/client.cc

```diff
@@ -854,6 +858,167 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+    }
+
+    static size_t calc_part_size(size_t total_size, size_t maximum_part_size) {
+        assert(maximum_part_size >= aws_minimum_part_size);
+        for (size_t part_size = maximum_part_size; part_size >= aws_minimum_part_size; part_size -= 1_MiB) {
+            size_t num_parts = (total_size + part_size) / part_size;
+            if (num_parts <= aws_maximum_parts_in_piece) {
+                return part_size;
+            }
+        }
+        s3l.error("unable to figure out part_size: total_size={}, maximum_part_size={}",
+                  total_size, maximum_part_size);
+        throw std::domain_error(fmt::format("total_size={}, maximum_part_size={}",
+                                            total_size, maximum_part_size));
+    }
+
+    future<> multipart_upload() {
+        // a background job which consumes the buffers pushed to part_body,
+        // and write them to with multipart upload
+        co_await start_upload();
+        size_t part_size = 0;
+        for (size_t offset = 0; offset < _total_size; offset += part_size) {
+            part_size = std::min(_total_size - offset, _part_size);
+            co_await upload_part(offset, part_size);
+        }
+        co_await finalize_upload();
+    }
+
+public:
+    upload_ranged_sink(shared_ptr<client> cln,
+                       sstring object_name,
+                       std::optional<tag> tag,
+                       size_t total_size,
+                       size_t max_part_size)
+        : upload_sink_base{std::move(cln), std::move(object_name), std::move(tag)}
+        , _part_size{calc_part_size(total_size, max_part_size)}
+        , _total_size{total_size} {
+        assert(_part_size >= aws_minimum_part_size);
+    }
+
+    future<> put(temporary_buffer<char> buf) final {
+        if (!_uploading) {
+            _uploading = multipart_upload();
+        }
+        co_await _part_body.writer.write(std::move(buf));
+    }
+
+    future<> put(net::packet packet) final {
+        for (auto&& buffer : packet.release()) {
+            co_await put(std::move(buffer));
+        }
+    }
+
+    future<> flush() final {
+        if (auto uploading = std::exchange(_uploading, {}); uploading) {
+            return *std::move(uploading);
+        }
+        return make_ready_future();
+    }
+
+    future<> close() final {
+        if (auto uploading = std::exchange(_uploading, {}); uploading) {
+            // close from the writer side
+            std::ignore = std::move(_part_body.writer);
+            return *std::move(uploading);
+        }
+        return make_ready_future();
+    }
+};
+
+data_sink client::make_upload_ranged(sstring object_name,
+                                     size_t total_size,
+                                     std::optional<tag> tag,
+                                     std::optional<size_t> max_part_size) {
+    assert(total_size > 0);
+    size_t maximum_part_size = max_part_size.value_or(aws_minimum_part_size);
```

Comment:
If a caller passes max_part_size disengaged, then it likely expects that the sink would pick the part size on its own. However, the calc_part_size() will effectively try to use aws minimal part size _only_. If total_size >= aws_minimal_part_size * aws_maximum_parts_in_piece` it throws.

Reply:
yes, the algorithm is buggy. i re-implemented it.